presto-main/src/main/java/com/facebook/presto/operator/aggregation/OptimizedTypedSet.java

```diff
@@ -0,0 +1,355 @@
+    private final int hashMask;
+
+    private int size;  // size is the number of elements added to the TypedSet (including null).
+    private Block[] blocks;   // Keeps track of the added blocks, even if the elements of the block was not inserted into the set. Array is used to get higher performance in getInsertPosition()
+    private List<SelectedPositions> positionsForBlocks;  // The selected positions for the added blocks, one for each added block
+    private long[] blockPositionByHash;  // Each 64-bit long is 32-bit index for blocks + 32-bit position within block
+    private int currentBlockIndex = -1;  // The index into the blocks array and positionsForBlocks list
+
+    public OptimizedTypedSet(Type elementType, int maxPositionCount, String functionName)
+    {
+        this(elementType, INITIAL_BLOCK_COUNT, maxPositionCount, functionName);
+    }
+
+    public OptimizedTypedSet(Type elementType, int expectedBlockCount, int maxPositionCount, String functionName)
+    {
+        checkArgument(expectedBlockCount >= 0, "expectedBlockCount must not be negative");
+        checkArgument(maxPositionCount >= 0, "maxPositionCount must not be negative");
+
+        this.elementType = requireNonNull(elementType, "elementType must not be null");
+        this.functionName = functionName;
+        this.hashCapacity = arraySize(maxPositionCount, FILL_RATIO);
+        this.hashMask = hashCapacity - 1;
+
+        blocks = new Block[expectedBlockCount];
+        positionsForBlocks = new ArrayList<>(expectedBlockCount);
+        blockPositionByHash = initializeHashTable();
+    }
+
+    /**
+     * Union the set by adding the elements of the block, eliminating duplicates.
+     */
+    public void union(Block block)
+    {
+        int positionCount = block.getPositionCount();
+
+        currentBlockIndex++;
+        ensureBlocksCapacity(currentBlockIndex + 1);
+        blocks[currentBlockIndex] = block;
+
+        int[] positions = new int[positionCount];
+
+        // Add the elements to the hash table. Since union can only increase the set size, there is no need to create a separate hashtable.
+        int positionsIndex = 0;
+        for (int i = 0; i < positionCount; i++) {
+            int hashPosition = getInsertPosition(blockPositionByHash, getMaskedHash(hashPosition(elementType, block, i)), block, i);
+            if (hashPosition != INVALID_POSITION) {
+                // There is no need to test if adding element is successful since it's on the same hash table
+                addElement(blockPositionByHash, hashPosition, block, i);
+                positions[positionsIndex++] = i;
+            }
+        }
+
+        getPositionsForBlocks().add(positionsList(positions, 0, positionsIndex));
+        size += positionsIndex;
+    }
+
+    /**
+     * Intersect the set with the elements of the block, eliminating duplicates.
+     */
+    public void intersect(Block block)
+    {
+        int positionCount = block.getPositionCount();
```

Comment:
let's move this inside the ```else``` to keep the scope small

Reply:
perhaps we don't even need this variable. Can inline it into the loop directly

Reply:
> perhaps we don't even need this variable. Can inline it into the loop directly

`positionCount` is used in the performance critical `for` loop. I'm sure the function call will be inlined here, but it may still be an indirect memory access through the `block` pointer. Besides, it's used in two different places and having a variable doesn't reduce readability. I'd prefer keeping the variable. I did move the declaration to inside of the `else` block.